File: app/models/World.scala
package models

import java.util.Set;
import org.neo4j.graphdb.Direction
import org.neo4j.graphdb.RelationshipType
import org.springframework.data.annotation.PersistenceConstructor
import models.annotations.graph_annotations._
import scala.collection.mutable._

@NodeEntity
case class World @PersistenceConstructor() (
                                             @Indexed var name: String,
                                             @Indexed var moons: Int
                                             ){
  @GraphId var id: java.lang.Long = _
  var spokenLanguage: String = _

  // TODO Polish on enums
  // TODO Does not receive any data
  @Fetch
  @RelatedTo(`type` = "REACHABLE_BY_ROCKET", direction = Direction.OUTGOING)
  var reachableByRocket2: scala.collection.mutable.Set[World] = _

  @Fetch
  @RelatedTo(`type` = "REACHABLE_BY_ROCKET", direction = Direction.OUTGOING)
  var reachableByRocket: Set[World] = _

}

object World {
  object RelTypes extends Enumeration {
      val REACHABLE_BY_ROCKET = new CustomRelationshipType("REACHABLE_BY_ROCKET")
      class CustomRelationshipType(val name: String) extends Val with RelationshipType
      implicit def convertValue(name: Value): CustomRelationshipType = name.asInstanceOf[CustomRelationshipType]
    }
}

File: app/models/World.java
package models;

import java.util.HashSet;
import java.util.Set;

import models.base.AbstractEntity;
import models.modelconstants.RelationshipTypesJava;
import org.neo4j.graphdb.Direction;
import org.springframework.data.neo4j.annotation.Fetch;
import org.springframework.data.neo4j.annotation.Indexed;
import org.springframework.data.neo4j.annotation.NodeEntity;
import org.springframework.data.neo4j.annotation.RelatedTo;
import org.springframework.data.neo4j.support.index.IndexType;

@NodeEntity
public class World extends AbstractEntity {

    @Indexed
    public String name;

    @Indexed(indexType = IndexType.LABEL)
    public int moons;

    public String spokenLanguage;

    @Fetch
    @RelatedTo(type = RelationshipTypesJava.REACHABLE_BY_ROCKET.Constant, direction = Direction.OUTGOING)
    public Set<World> reachableByRocket;

    // Constructors
    public World(String name, int moons, String spokenLanguage) {
        this.reachableByRocket = new HashSet<>();
        this.name = name;
        this.moons = moons;
        this.spokenLanguage = spokenLanguage;
    }

    public World(String name, int moons) {
        this.reachableByRocket = new HashSet<>();
        this.name = name;
        this.moons = moons;
    }

    private World() {
        this.reachableByRocket = new HashSet<>();
    }
}